builtin/credential/appgroup/path_tidy_user_id.go

```diff
@@ -0,0 +1,77 @@
+package appgroup
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathTidyUserID(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "tidy/user-id$",
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathTidyUserIDUpdate,
+		},
+
+		HelpSynopsis:    pathTidyUserIDSyn,
+		HelpDescription: pathTidyUserIDDesc,
+	}
+}
+
+// tidyUserID is used to delete entries in the whitelist that are expired.
+func (b *backend) tidyUserID(s logical.Storage) error {
+	grabbed := atomic.CompareAndSwapUint32(&b.tidyUserIDCASGuard, 0, 1)
+	if grabbed {
+		defer atomic.StoreUint32(&b.tidyUserIDCASGuard, 0)
+	} else {
+		return fmt.Errorf("user ID tidy operation already running")
+	}
+
+	userIDs, err := s.List("userid/")
+	if err != nil {
+		return err
+	}
+
+	for _, userID := range userIDs {
+		userIDEntry, err := s.Get("userID/" + userID)
+		if err != nil {
+			return fmt.Errorf("error fetching user ID %s: %s", userID, err)
+		}
+
+		if userIDEntry == nil {
```

Comment:
I wonder in these cases if we want to just clean out the user ID instead of erroring. Potentially add it as a `multierror` and keep going, or pass out a `logical.Response` and add it as a warning. The most harm that can happen is login is denied, so we fail secure. But otherwise I wonder if some bug could mean that someone could never get things clean.

(This is also a commentary on the other backends that tidy, too -- they should probably be consistent.)